ra-tls/libenclave-tls/enclave-tls.c

```diff
@@ -0,0 +1,295 @@
+{
+	(void) preverify;
+	int ret = 0;
+
+	// Need add oid in cert to identify the certificate type.
+	if (store->certs->length > 3000) {
+		void *handle = dlopen(LIBENCLAVE_TLS_SGX, RTLD_LAZY);
+		if (NULL == handle) {
+			fprintf(stderr, "open error:%s\n", dlerror());
+			return -1;
+		}
+		verify_cert_extensions_func verify_cert_extensions =
+			dlsym(handle, "verify_sgx_cert_extensions");
+		ret = verify_cert_extensions(store->certs->buffer,
+					     store->certs->length);
+		dlclose(handle);
+	}
+
+	fprintf(stderr, "Verifying certificate extensions ... %s\n",
+		ret == 0 ? "Success" : "Failure");
+	return !ret;
+}
+
+enclave_tls_ctx* enclave_tls_init(struct enclave_tls_conf *conf)
+{
+	enclave_tls_ctx *ctx = calloc(1, sizeof(*ctx));
+	if (ctx == NULL) {
+		fprintf(stderr, "ERROR: failed to calloc\n");
+		return NULL;
+	}
+	ctx->sockfd = conf->sockfd;
+	ctx->quote_type = conf->quote_type;
+	ctx->mutual = conf->mutual;
+	ctx->is_server = conf->is_server;
+
+	ctab.generate_certificate = &generate_x509;
+
+	if (ctx->quote_type == QUOTE_TYPE_EPID) {
+		ctx->handle = dlopen(LIBENCLAVE_TLS_SGX, RTLD_LAZY);
+		if (NULL == ctx->handle) {
+			printf("open error:%s\n", dlerror());
+			return -1;
+		}
+#define DLSYM(fn)									\
+		do {									\
+			ftab.fn = dlsym(ctx->handle, #fn "_sgx");				\
+		} while (0)
+
+		DLSYM(wolfSSL_free);
+		DLSYM(wolfSSL_CTX_set_verify);
+		DLSYM(create_key_and_x509);
+		DLSYM(wolfSSL_read);
+		DLSYM(wolfSSL_write);
+		DLSYM(wolfSSL_CTX_new);
+		DLSYM(printf_identity);
+		DLSYM(wolfTLSv1_2_client_method);
+		DLSYM(wolfSSL_X509_get_der);
+		DLSYM(wolfSSL_get_peer_certificate);
+		DLSYM(wolfSSL_connect);
+		DLSYM(wolfSSL_negotiate);
+		DLSYM(wolfSSL_set_fd);
+		DLSYM(wolfSSL_new);
+		DLSYM(wolfTLSv1_2_server_method);
+		DLSYM(wolfSSL_CTX_free);
+		DLSYM(wolfSSL_Cleanup);
+		DLSYM(wolfSSL_Debugging_ON);
+		DLSYM(wolfSSL_Debugging_OFF);
+		DLSYM(wolfSSL_Init);
+#undef DLSYM
+	} else {
+#define DLSYM(fn)                                                                       \
+                do {                                                                    \
+                        ftab.fn = &fn;                                                  \
+                } while (0)
+
+		DLSYM(wolfSSL_free);
+		DLSYM(wolfSSL_CTX_set_verify);
+		DLSYM(create_key_and_x509);
+		DLSYM(wolfSSL_read);
+		DLSYM(wolfSSL_write);
+		DLSYM(wolfSSL_CTX_new);
+		DLSYM(printf_identity);
+		DLSYM(wolfTLSv1_2_client_method);
+		DLSYM(wolfSSL_X509_get_der);
+		DLSYM(wolfSSL_get_peer_certificate);
+		DLSYM(wolfSSL_connect);
+		DLSYM(wolfSSL_negotiate);
+		DLSYM(wolfSSL_set_fd);
+		DLSYM(wolfSSL_new);
+		DLSYM(wolfTLSv1_2_server_method);
+		DLSYM(wolfSSL_CTX_free);
+		DLSYM(wolfSSL_Cleanup);
+		DLSYM(wolfSSL_Debugging_ON);
+		DLSYM(wolfSSL_Debugging_OFF);
+		DLSYM(wolfSSL_Init);
+#undef DLSYM
+	}
+
+#ifdef DEBUG
+	ftab.wolfSSL_Debugging_ON();
+#else
+	ftab.wolfSSL_Debugging_OFF();
+#endif
```

Comment:
it is better if debug on or off can be determined by runtime instead of build time.